### PR TITLE
AggregationLayer : Improve aggregation state management

### PR DESCRIPTION
**Pull request by 1chandu**
For #3844
Different approach than : #3850
- Add support for `dimensions`
- Add support for detecting when positions changed or any attribute is changed.
- Add support for detecting when extension props changed.
- Update documents, and unit tests.

#### Change List
- AggregationLayer: add support for aggregation change detection.

### CODE REVIEW

File: test/modules/aggregation-layers/aggregation-layer.spec.js [hunk 1]
```diff
@@ -55,17 +59,23 @@ class TestAggregationLayer extends AggregationLayer {
     // clear state
     this.setState({aggregationDirty: false});
     super.updateState(opts);
-    this.setState({aggregationDirty: this.isAggregationDirty(opts)});
+    this.setState({
+      aggregationDirty: this.isAggregationDirty(opts, {
+        dimension: this.state.dimensions.data,
+        detectExtensionChange: true
```
Comment (1chandu) [inline test/modules/aggregation-layers/aggregation-layer.spec.js]: rename, compareAll ?

File: modules/aggregation-layers/src/aggregation-layer.js [hunk 2]
```diff
@@ -86,20 +91,47 @@ export default class AggregationLayer extends CompositeLayer {
...
+    const {props, oldProps, changeFlags} = updateOpts;
+    const {detectExtensionChange = false, dimension} = params;
+    const {ignoreProps} = this.state;
+    const {props: dataProps, accessors} = dimension;
+    const {updateTriggersChanged} = changeFlags;
+    if (updateTriggersChanged) {
+      if (updateTriggersChanged.all) {
+        return true;
+      }
+      if (accessors.some(accessor => updateTriggersChanged[accessor])) {
```
Comment (1chandu) [inline modules/aggregation-layers/src/aggregation-layer.js]: change to for loop

File: modules/aggregation-layers/src/utils/prop-utils.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+export function filterProps(props, filterKeys) {
+  return Object.keys(props).reduce((accu, cur) => {
```
Comment (1chandu) [inline modules/aggregation-layers/src/utils/prop-utils.js]: For loop

File: modules/aggregation-layers/src/aggregation-layer.js [hunk 4]
```diff
@@ -18,24 +18,25 @@
...
-  initializeState(aggregationProps = []) {
+  initializeState(dimensions) {
     super.initializeState();
+
+    // props , when changed doesn't require updating aggregation
+    const ignoreProps = filterProps(this.constructor._propTypes, dimensions.data.props);
+
     this.setState({
-      aggregationProps: aggregationProps.concat(UNIFORM_PROPS)
+      ignoreProps: Object.assign(ignoreProps, Layer.defaultProps),
```
Comment (1chandu) [inline modules/aggregation-layers/src/aggregation-layer.js]: Layer.defaultProps not needed.

Comment (1chandu) [reply modules/aggregation-layers/src/aggregation-layer.js]: If updateTrigger[key] and key is not in `ignoreProps`  - trigger re-aggreagtion (for GPU only). => attributeChanged is enough.

File: modules/aggregation-layers/src/aggregation-layer.md [hunk 5]
```diff
@@ -22,14 +22,43 @@ attributeManager.add({
...
+This helper can be used if a dimension is changed. Sublayers can defined custom dimensions and call this method to check if a dimension is changed.
+
+
+### isAttributeChanged()
+
+`AggregationLayer` tracks what attributes are changed in each update cycle. Super classes can use `isAttributeChanged()` method to check if a specific attribute is changed or any attribute is changed.
+
+#### Aggregation State
+
+`AggregationLayer` is responsible for setting following values in `sate` object:
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/aggregation-layer.md]: Looks like these are replaced by `isAttributeChanged()`?

File: modules/aggregation-layers/src/utils/grid-aggregation-utils.js [hunk 6]
```diff
@@ -35,6 +38,27 @@ export function getBoundingBox(attributes, vertexCount) {
 }
 /* eslint-enable max-statements */
 
+export function alignBoundingBox(boundingBox, gridOffset, coordinateSystem, viewport) {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/utils/grid-aggregation-utils.js]: How come the cell size is named `gridOffset`?

File: test/browser.js [hunk 7]
```diff
@@ -49,7 +49,7 @@ test('deck.gl', t => {
   require('./modules/main/bundle');
   require('./modules/aggregation-layers/utils/gpu-grid-aggregator.spec');
   require('./modules/aggregation-layers/gpu-grid-layer/gpu-grid-layer.spec');
-  require('./modules/aggregation-layers/grid-aggregation-layer.spec');
+  // require('./modules/aggregation-layers/grid-aggregation-layer.spec');
```
Comment (Pessimistress) [inline test/browser.js]: Restore?

File: modules/aggregation-layers/src/utils/grid-aggregation-utils.js (hunk 6 above)
Comment (1chandu) [reply modules/aggregation-layers/src/utils/grid-aggregation-utils.js]: This is how it is in CPU aggregation code (hexagon and cpugridlayer), there are too many references of `gridOffset`, will make the name change in a different PR.

File: test/browser.js (hunk 7 above)
Comment (1chandu) [reply test/browser.js]: Addressing this in #4046

**Pull request merged**